### PR TITLE
Code share page

### PR DESCRIPTION
**Pull request by rjvelazco**

### CODE REVIEW

File: components/CodeSharePost.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import Link from 'next/link';
+
+// helpers
+import { getDate } from '@helpers/data-formatter';
+
+// Models
+import { codesharePost } from '@models/CodeShare.model';
```
Comment (fmontes) [inline components/CodeSharePost.tsx]:
```suggestion
import { CodeSharePost } from '@models/CodeShare.model';
```

Interfaces should be TitleClase for convention.

File: components/header/components/HeaderNav.tsx [hunk 2]
```diff
@@ -10,7 +10,7 @@ export const HeaderNav = (): JSX.Element => {
             id: 'realease',
             link: '/latest/current-releases'
         },
-        { title: 'Code Share', id: 'code_share', link: '/' },
+        { title: 'Code Share', id: 'code_share', link: '/codeshare/all/1' },
```
Comment (fmontes) [inline components/header/components/HeaderNav.tsx]: Can we do this just `/codeshare/`?

File: helpers/data-formatter.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+export const getDate = (date: string): string => {
+    const created = new Date(date);
+    return `${created.toLocaleDateString('en-US', { month: 'short' })} ${created.getDay()}, 
+    ${created.getFullYear()}`;
+};
+
+export const printTags = (tags: string[]): string[] => {
+    const lastIndet = tags.length - 1;
```
Comment (fmontes) [inline helpers/data-formatter.ts]:
```suggestion
    const lastIndex = tags.length - 1;
```

File: helpers/data-formatter.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+export const getDate = (date: string): string => {
+    const created = new Date(date);
+    return `${created.toLocaleDateString('en-US', { month: 'short' })} ${created.getDay()}, 
+    ${created.getFullYear()}`;
+};
+
+export const printTags = (tags: string[]): string[] => {
```
Comment (fmontes) [inline helpers/data-formatter.ts]: What this functions does?

File: models/CodeShare.model.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+export interface codesharePost {
+    authorName: string;
+    dateCreated: string;
+    title: string;
+    urlTitle: string;
+    seoDescription: string;
+}
+
+export interface codesharePage extends codesharePost {
```
Comment (fmontes) [inline models/CodeShare.model.ts]:
```suggestion
export interface CodeSharePage extends codesharePost {
```

File: models/CodeShare.model.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+export interface codesharePost {
```
Comment (fmontes) [inline models/CodeShare.model.ts]:
```suggestion
export interface CodeSharePost {
```

File: pages/_app.tsx [hunk 7]
```diff
@@ -31,14 +34,15 @@ function MyApp({ Component, pageProps }: AppProps): JSX.Element {
...
+                <title>Documentation</title>
                 <meta
                     content="width=device-width, initial-scale=1.0, user-scalable=0"
                     name="viewport"
                 />
             </Head>
             <GlobalStyle />
             {pageProps.navDot ? (
-                <Grid>
+                <Grid codeShare={false}>
```
Comment (fmontes) [inline pages/_app.tsx]: Ugghhhh not fan of this.

File: pages/codeshare/[tag]/[pag].tsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+interface pageProps {
+    data: codesharePost[];
+    page: number;
+    totalCount: number;
+    tag: string;
+}
+
+const postPerPage = 10;
+
+export default function Home({ data, page, totalCount, tag }: pageProps): JSX.Element {
```
Comment (fmontes) [inline pages/codeshare/[tag]/[pag].tsx]:
```suggestion
export default function CodeShareTag({ data, page, totalCount, tag }: pageProps): JSX.Element {
```

File: pages/codeshare/[tag]/[pag].tsx [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+// Graphql
+import { CODE_SHARE_QUERY_LIST_ARTICULES, CODE_SHARE_QUERY_TOTAL_COUNT } from '@graphql/queries';
+
+// Utils
+import { client } from '@utils/graphql-client';
+
+// Models
+import { codesharePost } from '@models/CodeShare.model';
+
+interface pageProps {
```
Comment (fmontes) [inline pages/codeshare/[tag]/[pag].tsx]:
```suggestion
interface PageProps {
```

File: pages/codeshare/[tag]/[pag].tsx [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        'py',
+        'rounded',
+        'focus:outline-none'
+    ];
+    const totalCountPage = page * postPerPage;
+    return (
+        <>
+            {page > 1 ? (
+                <Link href={`/codeshare/${urlTag}${page - 1}`}>
+                    <button className={classNames(buttonClasses)}>Previous</button>
```
Comment (fmontes) [inline pages/codeshare/[tag]/[pag].tsx]: `<button>` for actions `<a>` for navigation.

File: pages/codeshare/post/[urlTitle].tsx [hunk 11]
```diff
@@ -0,0 +1,121 @@
+import React from 'react';
```
Comment (fmontes) [inline pages/codeshare/post/[urlTitle].tsx]: We need to remove the `post` section here.

We need the URL to be the same as now.

File: plugins/DotCodeMultiline.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    return function (node: CustomNode): void {
+        visit(node, 'jsx', (node: CustomNode) => {
+            const value = node.value;
+            const preContent = new RegExp(/<pre>(.|\n)*?<\/pre>/gi);
+            node.value = value.replace(preContent, (match) => {
+                return match.replace(/\n/g, '<br />');
+            });
+        });
+    };
+}
```
Comment (fmontes) [inline plugins/DotCodeMultiline.ts:19]: What is this for?

**Review by fmontes — CHANGES_REQUESTED** [review]

File: helpers/data-formatter.ts (hunk 4 above)
Comment (rjvelazco) [reply helpers/data-formatter.ts]: pringTag is a function that helps to format the array tag of each post in an ordered string. However, as there is a problem with the field is in graphql, I am not using it.

File: plugins/DotCodeMultiline.ts (hunk 12 above)
Comment (rjvelazco) [reply plugins/DotCodeMultiline.ts:19]: That's the plugin to solve the issue that `<pre>` and `<code>` tags had, which made all the code remain on one line.

File: components/CodeSharePost.tsx [hunk 13]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import Link from 'next/link';
+
+// helpers
+import { getDate } from '@helpers/data-formatter';
+
+// Models
+import { CodeSharePostInterface } from '@models/CodeShare.model';
```
Comment (fmontes) [inline components/CodeSharePost.tsx]:
```suggestion
import { CodeShareItem } from '@models/CodeShare.model';
```

Don't use the word `Interface` or `Model` is redundant (unless is really necessary)

PS. Naming stuff is hard 😂

File: components/CodeSharePost.tsx [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+export const CodeSharePost = ({ data }: { data: CodeSharePostInterface }): JSX.Element => {
+    return (
+        <div className="mb-14 pr-10">
+            <Link href={`/codeshare/${data.urlTitle}`}>
+                <a className="font-bold inline-block mb-1 no-underline text-blue-500 text-lg hover:underline">
+                    <h3 className="m-0">{data.title}</h3>
+                </a>
+            </Link>
+            <p className="mb-1">
+                <span className="font-bold">Created:</span> {getDate(data.dateCreated)}
```
Comment (fmontes) [inline components/CodeSharePost.tsx]: Let's be semantic here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time

Also, you might want to create a component to display dates.

File: components/CodeSharePost.tsx [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+            </p>
+            {data.authorName ? (
+                <p className="mb-1">
+                    <span className="font-bold">Author:</span> {data.authorName}
+                </p>
+            ) : null}
+            <p className="text-lg leading-7">{data.seoDescription}</p>
```
Comment (fmontes) [inline components/CodeSharePost.tsx]: Are you sure the field is `seoDescription`?

File: components/CodeShareSide.tsx [hunk 16]
```diff
@@ -0,0 +1,39 @@
+import React, { ReactNode } from 'react';
+import classNames from 'classnames';
+
+export const CodeShareSide = ({ children }: { children: ReactNode }): JSX.Element => {
+    return (
+        <div className="w-72">
```
Comment (fmontes) [inline components/CodeShareSide.tsx]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/aside

File: components/CodeShareSide.tsx [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+        'text-blue-300'
+    ];
+    return (
+        <div className="bg-blue-100 p-4 px-5 rounded w-full">
+            <h3 className="m-0 mb-8 text-blue-400 text-center">
+                Have you created code you&apos;d like to share with the dotCMS community?
+            </h3>
+            <div className="text-sm">
+                <a href="https://dotcms.com/codeshare/submit-code">
+                    <button className={classNames(btnCodeShare)}>SHARE YOUR CODE</button>
```
Comment (fmontes) [inline components/CodeShareSide.tsx]: You shouldn't have a button inside a `<a>` is not valid HTML5 code.

If you need the element to look like a button, just style it as a button.

File: components/CodeShareTopics.tsx [hunk 18]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import Link from 'next/link';
+
+// Interfaces
+import { CodeShareTopicsInterface } from '@models/CodeShare.model';
```
Comment (fmontes) [inline components/CodeShareTopics.tsx]:
```suggestion
import { CodeShareTopic } from '@models/CodeShare.model';
```

A interface represent an instance of ONE element.

File: components/CodeSharePost.tsx [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+import { CodeSharePostInterface } from '@models/CodeShare.model';
+
+export const CodeSharePost = ({ data }: { data: CodeSharePostInterface }): JSX.Element => {
+    return (
+        <div className="mb-14 pr-10">
+            <Link href={`/codeshare/${data.urlTitle}`}>
+                <a className="font-bold inline-block mb-1 no-underline text-blue-500 text-lg hover:underline">
+                    <h3 className="m-0">{data.title}</h3>
+                </a>
+            </Link>
```
Comment (fmontes) [inline components/CodeSharePost.tsx]:
```suggestion
        <h3 className="m-0">
            <Link href={`/codeshare/${data.urlTitle}`}>
                <a className="font-bold inline-block mb-1 no-underline text-blue-500 text-lg hover:underline">
                    {data.title}
                </a>
            </Link>
        </h3>
```

You can't have an h3 inside an a is not valid HTML 5.

File: components/CodeSharePost.tsx (hunk 14 above)
Comment (fmontes) [inline components/CodeSharePost.tsx]:
```suggestion
                <b>Created:</b> {getDate(data.dateCreated)}
```

File: components/CodeSharePost.tsx (hunk 15 above)
Comment (fmontes) [reply components/CodeSharePost.tsx]: `leading-7`? why don't just use margin?

File: components/CodeShareSide.tsx [hunk 20]
```diff
@@ -0,0 +1,39 @@
+import React, { ReactNode } from 'react';
+import classNames from 'classnames';
+
+export const CodeShareSide = ({ children }: { children: ReactNode }): JSX.Element => {
+    return (
+        <div className="w-72">
+            <h3 className="text-2xl">Topics</h3>
+            {children}
+            <SideShareCode />
```
Comment (fmontes) [inline components/CodeShareSide.tsx]: `SideShareCode` should be a children too.

File: components/CodeShareTopics.tsx [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+    topics
+}: {
+    topics: CodeShareTopicsInterface[];
+}): JSX.Element => {
+    return (
+        <ul className="list-none">
+            {topics.map((topic, index) => (
+                <li key={index}>
+                    <Link href={`/codeshare/topic/${topic.link}/1`}>
+                        <a className="cursor-pointer no-underline text-blue-500">{topic.title}</a>
```
Comment (fmontes) [inline components/CodeShareTopics.tsx]: `cursor-pointer` should be in all `<a>` no need to add it to each.

File: pages/_app.tsx [hunk 22]
```diff
@@ -31,6 +34,7 @@ function MyApp({ Component, pageProps }: AppProps): JSX.Element {
     return (
         <>
             <Head>
+                <title>Documentation</title>
```
Comment (fmontes) [inline pages/_app.tsx]: Are we passing this page based?

File: pages/codeshare/[urlTitle].tsx [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+export default function CodeShare({ data, source, error }: pageData): JSX.Element {
+    const content = source ? hydrate(source) : null;
+    const mainClasses = ['container', 'mx-auto', 'px-16', 'flex-grow'];
+    return (
+        <>
+            {error ? (
+                <main className={classNames(mainClasses)}>
+                    <h1>{data.title}</h1>
+                    <Terminal content={error} />
+                </main>
```
Comment (fmontes) [inline pages/codeshare/[urlTitle].tsx]: Extract this to a component that we can reuse.

**Review by fmontes — CHANGES_REQUESTED** [review]

**Pull request merged**